### ISSUE TITLE
Wrap Eidos code edition/analysis features into a proper language server.

### ISSUE THREAD

**Issue by iago-lito** [MesserLab/SLiM#424]
Just tossing the idea here for future reference, in case anyone finds the ressource to tackle it.

The Eidos "input commands" subwindow of the SLiMGui offers sophisticated code-edition-related features like:
- syntax highlighting
- autoformatting
- autocompletion
- errors linting
- code navigation
- ..

These must be the result of a huge work, and they make the Eidos coding experience rather nice. Unfortunately, they are restricted to this very particular sub-window of this very particular SLiMGui application. As a consequence, while writing Eidos code, we cannot benefit from any other standard modern IDE features like shortcuts customization, macros recording, automation, multiple window management, split views, theming, _etc._ unless we use an external editor instead of the SLiMGui editor. But then we loose the above features.

If the above features were extracted into a proper `eidos-`[`language-server`](https://langserver.org/), then we could have our cake and eat it too. We could edit Eidos script files using either SLiMGui/vim/vscode/intellij/emacs/atom/notepad++/\<whatever\>.. and still always benefit from the above features.

**Comment by bhaller** [comment]
As discussed over email with Iago, I've marked this "help wanted"; it might be a fun side project for somebody, but will be complex, and I'm unlikely to have time for it for, well, maybe ever.  But it's a nice idea.

**Comment by csmcal** [comment]
Hi all, I'm joining from the google discussion group thread. I had been working on a vscode extension of my own (https://github.com/csmcal/slim-eidos-vsc-extension), but my focus had not been on making a feature complete language server. I do see the benefit though.

@andrewkern I have some type implementation to allow some object oriented documentation functionality, but I'm new to this and not really sure what to do from there.

For the .svg icon, my reading of the note in the [Publication Guidelines](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#publishing-extensions) is that using such will prevent it from being published. I'm not sure if this is a policy that has changed or will changed, or that I am interpreting it correctly, but that is what gave me the idea.

@bhaller my main thought of the benefit of your collaboration was that we might have an automated workflow to update the language server alongside SLiM updates when provided (by you) with an updated simple documentation. Something simpler to programmatically read than the ref sheet pdfs, though perhaps something is already available somewhere?

Maybe we could make a github organization for SLiM community repositories and projects. Joint ownership and passing ownership might give some future-proofing to maintenance, and such an org would give a place to direct people to host their community project ideas. Thoughts?

**Comment by andrewkern** [comment]
> Maybe we could make a github organization for SLiM community repositories and projects. Joint ownership and passing ownership might give some future-proofing to maintenance, and such an org would give a place to direct people to host their community project ideas. Thoughts?

I agree with this. @bhaller having an organization could mean that tools related to SLiM could all live in one place on Github, similar to what we've done with [stdpopsim](https://github.com/popsim-consortium) or the even more mature [tskit-dev](https://github.com/tskit-dev). @petrelharp what would you think of a SLiM GH org for putting tools up?

**Comment by andrewkern** [comment]
> [@andrewkern](https://github.com/andrewkern) I have some type implementation to allow some object oriented documentation functionality, but I'm new to this and not really sure what to do from there.
> 
> For the .svg icon, my reading of the note in the [Publication Guidelines](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#publishing-extensions) is that using such will prevent it from being published. I'm not sure if this is a policy that has changed or will changed, or that I am interpreting it correctly, but that is what gave me the idea.

thanks for this info on the `.svg` @csmcal. I'm going to tag you on an issue on my repo to continue the conversation on OO docs-- i've made a lot of progress on this over the past two days

**Comment by bhaller** [comment]
> > Maybe we could make a github organization for SLiM community repositories and projects. Joint ownership and passing ownership might give some future-proofing to maintenance, and such an org would give a place to direct people to host their community project ideas. Thoughts?
> 
> I agree with this. [@bhaller](https://github.com/bhaller) having an organization could mean that tools related to SLiM could all live in one place on Github, similar to what we've done with [stdpopsim](https://github.com/popsim-consortium) or the even more mature [tskit-dev](https://github.com/tskit-dev). [@petrelharp](https://github.com/petrelharp) what would you think of a SLiM GH org for putting tools up?

Discussed this a bit with @andrewkern on Slack.  A difference from stdpopsim or tskit is that SLiM itself lives inside the MesserLab org, and is not likely to move any time soon.  So what we're talking about here is making an org that is for the "SLiM ecosystem", but not for SLiM itself.  Thus far, I've been putting that kind of stuff in the SLiM-Extras repo, which is also inside MesserLab.  So I see a couple of options here:

- You guys could put your stuff into SLiM-Extras, if it suits your purposes.  I'd be happy to give you all permissions to muck about with that repo, so that I don't need to be a middleman, approve PRs, etc.; you'd be in charge.  Your stuff would presumably go into new folder(s) within that repo.
- But maybe it makes more sense to keep SLiM-Extras as a repo for actual `.slim` models and such – extra script examples and recipes and solutions for SLiM users.  Maybe a new org makes sense as a place for SLiM-related _tools_, which is a pretty different and new concept.  If so, great!  I'm happy to make a new org, with whatever name you guys want; just tell me what to do.
- But maybe it makes more sense for you guys to create this org yourselves, and be in charge of it.  I don't think I provide any value-add here really.  Maybe it makes it look more "officially SLiM" if it has my name on it?  I dunno.  Whatever you want.  To me, it seems better for you to do it, because then it's really 100% yours; I shouldn't take credit for things I haven't done!  :->

The ball is now in your court.  :->  Besides the people already tagged on this issue, perhaps @philippmesser and @petrelharp would be interested in this discussion.  If you folks give me an action item, I'll do it.  Besides that, I'll just say I'm excited about this stuff, and looking forward to seeing it out in the real world!  :->

**Comment by andrewkern** [comment]
Yeah @petrelharp -- what do you think of a `SLiM-Community-Tools` Organziation on GH as a clearing house for tools we develop on the back of SLiM? I'm thinking things like [slim_magic](https://github.com/andrewkern/slim_magic) could go there?

**Comment by petrelharp** [comment]
I think `slim_community` would be a more concise name? I feel like the missing part to that would be where does that community communicate, though - if there's no place for communication and cohesion then there's no community / organization. There could be a github-based medium for discussion? Or a SLiM slack?

**Comment by bhaller** [comment]
GitHub does have Discussions; I haven't used it much, and am not sure how well it works for this sort of thing.  A SLiM slack is certainly possible; seems like it might overlap a lot with slim-discuss, and the line between those two things might be rather blurry.  I'm not sure.  Maybe if we were firm from the beginning: slim-discuss is for asking questions and getting help, and if someone tries to use the Slack for that, we redirect them to slim-discuss?  (The virtue of slim-discuss being that it is searchable through Google, provides a separate thread for each question, etc.)  Or maybe slim-discuss would get phased out in favor of the Slack?  I'm not good at these kinds of decisions, really, I don't think.

**Comment by csmcal** [comment]
@petrelharp A fan of `_` more than `-`? `slim_community` is more concise, and I favor short repo addresses. 

As @bhaller pointed out while I was typing this Github Orgs can enable the Github Discussions feature, which makes something of a bbs under a tab along the top like the 'Issues' tab, you can see the one enabled for `MesserLab/SLiM` The Discussions feature for an org is housed in an extra repo under the org, I would happily set one up. I imagine part of the discussion will also be interactive with the google discussion group for SLiM, and I think if we use the github tab we might as well just keep it to slim-discuss and the github org and not open a slack.

@andrewkern did you tag an issue? I checked but may have missed it this week. Happy to discuss, though reading a lot more on LSPs I don't think I'll have the bandwidth to contribute a major portion of one, at least beyond rewrites or integration of what I've done.

@bhaller orgs can have co-owners, and I think it would be useful to give you such authority so that you might add people interested in collaborating in the future, if they had reached out to you with interest in something housed there

**Comment by csmcal** [comment]
....and now I assume you're all familiar with the discussion feature as you've all separately posted there for this repo in the past :p

**Comment by bhaller** [comment]
> ....and now I assume you're all familiar with the discussion feature as you've all separately posted there for this repo in the past :p

When someone starts a discussion, I respond to it, but I don't really understand the feature, what makes it different from issues, what it's capable of, how it would feel to try to have a community of users chatting using it, etc...

Anyhow, I like `slim_community` too.  @andrewkern do you like it?  If so, someone should make it (adding me as a co-owner, sure), and then @csmcal can set up the discussions feature, and then y'all can move repos into it (perhaps SLiM-Extras should be moved into it? I don't know how to do that, preserving the commit history and everything, but it seems appropriate, perhaps, since it is mostly community contributions?), and then we can announce it on slim-discuss and invite people to participate?  I'm on board, if everybody likes this idea!

We should think about what other stuff might move into it.  @samchamper has a small repo for generating plots of spatial models, IIRC?  @petrelharp you have some materials for teaching popgen using SLiM?  Are things like that appropriate for this org?  It would certainly make things like that easier for others in the community to find, if their repos moved into `slim_community".  But on the other hand, do we want a sudden explosion of dozens of small repos in the org?  Or is it intended more for larger projects like this VS code stuff?  This is another area where I feel like I don't have reliable instincts.

**Comment by andrewkern** [comment]
@csmcal -- see [here](https://github.com/andrewkern/slim-vscode-tools/issues/1). I made a lot of additions this week, and went live on the marketplace with a v0.0.2 release

@bhaller -- I like that name

**Comment by csmcal** [comment]
whoever CScam is just got pinged with something completely random 😂

**Comment by andrewkern** [comment]
Ooops!

**Comment by csmcal** [comment]
I can go ahead and make the org and add you as co-owners. `slim-community` to go with `slim-discuss`? @bhaller I think any community project is appropriate to include, as long as it is in relatively functional state or actively developed. Size shouldn't matter but more community networking is always good. Though, I'm also ambivalent on SLiM-Extras as that has material otherwise cut from the manual. GitHub has a guide on [transferring repositories](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository) we could direct people to.

**Comment by bhaller** [comment]
> I can go ahead and make the org and add you as co-owners. `slim-community` to go with `slim-discuss`?

I'd say go for it.  Seems like we have consensus.

> I think any community project is appropriate to include, as long as it is in relatively functional state or actively developed. Size shouldn't matter but more community networking is always good. Though, I'm also ambivalent on SLiM-Extras as that has material otherwise cut from the manual. GitHub has a guide on [transferring repositories](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#repository-transfers-and-organizations) we could direct people to.

OK, interesting.  Perhaps there are things within SLiM-Extras that ought to be moved over, such as @samchamper's sublaunching tutorial, and other things in SLiM-Extras that should stay there because they are more, as you say, leftovers cut from the manual and such.  Anyhow, those kinds of decisions can be made any time in the future, no urgency to them now.  For now, let's just get it set up with the stuff we know we want there, and get it looking nice for public consumption, and then announce it on slim-discuss.  Thanks!

**Comment by bhaller** [comment]
Joined.  Do you plan to make a README?  I could throw a draft together if you want me to?

**Comment by csmcal** [comment]
By all means, go ahead. I set up the discussions feature and am looking into it further atm

**Comment by bhaller** [comment]
OK, I just make a README.md.  I guess it is the "public" README?  It doesn't seem to display when I go to https://github.com/slim-community, though; not sure why.  I gather I could also make a README for "members", too, but maybe the one README can be used for both purposes?  I'm not sure the "members" one needs to be different.  But anyhow, it isn't displaying for me at all; the org site still just shows the "We think you’re gonna like it here" stuff.  Well, I just hid that; but still, the README I made is not showing.  Maybe one of you knows how to make it display.  :-O

**Comment by bhaller** [comment]
I just made a "members" README as well.  It's a very trimmed version of the "public" readme.  Seems reasonable to me, but feel free to edit what I've done.  I think I'm done fiddling for now, I've got to go do other things.  :->  But let's start throwing content on there!

**Comment by csmcal** [comment]
I can't find a way to share one readme for public/private, may not be possible. Thanks for the efforts! I was sidetracked in the access permissions

**Comment by bhaller** [comment]
> I can't find a way to share one readme for public/private, may not be possible. Thanks for the efforts! I was sidetracked in the access permissions. Continuing discussion within the org.

Yeah, I decided that I wanted the "public" and "members" READMEs to look different anyway, in the end.  :->

**Comment by csmcal** [comment]
Apologies if I jumped the gun in making the org as `slim-community`, I thought it was thematically consistent but didn't think the difference was critical. A general trend to `-` for thematic consistency might be nice but I see no reason to require repo name changes as part of an org style guide or similar.

I do think it is a good idea to check in with others before adding a repo, but I think it's also good to let members add repos that they feel make sense under their own power. We intend this to be an open community, so I think having low bounds by default should be the goal?

It might be worth adding org guidelines to the readme docs if any of you feel it worthwhile.

**Comment by bhaller** [comment]
> It might be worth adding org guidelines to the readme docs if any of you feel it worthwhile.

Maybe eventually we'll need to, but for now I think we're good.  :->  We can play it by ear, until it feels like that isn't working.
